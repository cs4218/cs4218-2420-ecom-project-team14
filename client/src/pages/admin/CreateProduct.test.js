// CreateProduct.test.js
import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { act } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";
import CreateProduct from "./CreateProduct";
import axios from "axios";
import toast from "react-hot-toast";

// Mock the dependencies
jest.mock("./../../components/Layout", () => {
  return ({ children, title }) => (
    <div data-testid="mock-layout" data-title={title}>
      {children}
    </div>
  );
});

jest.mock("./../../components/AdminMenu", () => {
  return () => <div data-testid="mock-adminmenu">Admin Menu</div>;
});

const mockedNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedNavigate,
}));

jest.mock("axios");
jest.mock("react-hot-toast");

global.URL.createObjectURL = jest.fn(() => "blob:http://localhost/dummy");

describe("CreateProduct Component", () => {
  beforeEach(() => {
    axios.get.mockResolvedValueOnce({
      data: { success: true, category: [] },
    });
    jest.clearAllMocks();
  });

  // Basic Rendering Tests
  describe("Component Rendering", () => {
    it("renders CreateProduct component without crashing", async () => {
      await act(async () => {
        render(<CreateProduct />);
      });
      expect(screen.getByTestId("mock-layout")).toBeInTheDocument();
    });

    it("renders with correct layout title", async () => {
      await act(async () => {
        render(<CreateProduct />);
      });
      expect(screen.getByTestId("mock-layout")).toHaveAttribute(
        "data-title",
        "Dashboard - Create Product"
      );
    });
  });

  // Structure Tests
  describe("Component Structure", () => {
    it("contains main container with bootstrap classes", async () => {
      let renderResult;
      await act(async () => {
        renderResult = render(<CreateProduct />);
      });
      const { container } = renderResult;
      const mainContainer = container.querySelector(".container-fluid");
      expect(mainContainer).toBeInTheDocument();
      expect(mainContainer).toHaveClass("m-3", "p-3");
    });

    it("contains two columns with correct classes", async () => {
      let renderResult;
      await act(async () => {
        renderResult = render(<CreateProduct />);
      });
      const { container } = renderResult;
      const col1 = container.querySelector(".col-md-3");
      const col2 = container.querySelector(".col-md-9");
      expect(col1).toBeInTheDocument();
      expect(col2).toBeInTheDocument();
    });
  });

  // Form Elements Tests
  describe("Form Elements", () => {
    it("renders form inputs and select elements", async () => {
      await act(async () => {
        render(<CreateProduct />);
      });

      expect(screen.getByText("Select a Category")).toBeInTheDocument();
      expect(screen.getByText("Upload Photo")).toBeInTheDocument();
      expect(screen.getByPlaceholderText("Write a Name")).toBeInTheDocument();
      expect(
        screen.getByPlaceholderText("Write a Description")
      ).toBeInTheDocument();
      expect(screen.getByPlaceholderText("Write a Price")).toBeInTheDocument();
      expect(
        screen.getByPlaceholderText("Write a Quantity")
      ).toBeInTheDocument();
      expect(screen.getByText("Select Shipping")).toBeInTheDocument();
      expect(screen.getByText("CREATE PRODUCT")).toBeInTheDocument();
    });

    it("handles input changes correctly", async () => {
      await act(async () => {
        render(<CreateProduct />);
      });

      const nameInput = screen.getByPlaceholderText("Write a Name");
      const descriptionInput = screen.getByPlaceholderText(
        "Write a Description"
      );
      const priceInput = screen.getByPlaceholderText("Write a Price");
      const quantityInput = screen.getByPlaceholderText("Write a Quantity");

      fireEvent.change(nameInput, { target: { value: "Test Product" } });
      fireEvent.change(descriptionInput, {
        target: { value: "Test Description" },
      });
      fireEvent.change(priceInput, { target: { value: "100" } });
      fireEvent.change(quantityInput, { target: { value: "10" } });

      expect(nameInput.value).toBe("Test Product");
      expect(descriptionInput.value).toBe("Test Description");
      expect(priceInput.value).toBe("100");
      expect(quantityInput.value).toBe("10");
    });
  });

  // Category Fetching Tests
  describe("Category Fetching", () => {
    it("fetches categories on component mount", async () => {
      await act(async () => {
        render(<CreateProduct />);
      });

      expect(axios.get).toHaveBeenCalledWith("/api/v1/category/get-category");
    });

    it("displays fetched categories in the select dropdown", async () => {
      const categories = [
        { _id: "1", name: "Category 1" },
        { _id: "2", name: "Category 2" },
      ];

      axios.get.mockReset();
      axios.get.mockResolvedValueOnce({
        data: { success: true, category: categories },
      });

      await act(async () => {
        render(<CreateProduct />);
      });

      expect(axios.get).toHaveBeenCalledWith("/api/v1/category/get-category");
    });

    it("shows toast error if category fetching fails", async () => {
      axios.get.mockReset();
      axios.get.mockRejectedValueOnce(new Error("Network Error"));

      await act(async () => {
        render(<CreateProduct />);
      });

      expect(toast.error).toHaveBeenCalledWith(
        "Something went wrong in getting category"
      );
    });
  });

  // File Upload Tests
  describe("File Upload", () => {
    it("handles file upload and displays preview", async () => {
      await act(async () => {
        render(<CreateProduct />);
      });

      const file = new File(["dummy content"], "test.png", {
        type: "image/png",
      });

      const uploadLabel = screen.getByText("Upload Photo");
      const fileInput = uploadLabel.querySelector('input[type="file"]');

      fireEvent.change(fileInput, { target: { files: [file] } });

      await waitFor(() => {
        const previewImage = screen.getByAltText("product_photo");
        expect(previewImage).toBeInTheDocument();
        expect(previewImage).toHaveAttribute(
          "src",
          "blob:http://localhost/dummy"
        );
      });
    });
  });

  // Form Submission Tests
  describe("Form Submission", () => {
    it("submits form with correct data", async () => {
      jest.mock("antd", () => {
        // Create simple Select mock that don't try to replicate behavior
        const MockSelect = ({ children, onChange, placeholder }) => (
          <div data-testid={`mock-select-${placeholder}`}>{children}</div>
        );

        MockSelect.Option = ({ children, value }) => (
          <div data-testid={`mock-option-${value}`}>{children}</div>
        );

        return {
          Select: MockSelect,
        };
      });
      // Mock API responses
      axios.post.mockResolvedValueOnce({
        data: { success: true, message: "Product Created Successfully" },
      });

      axios.get.mockResolvedValueOnce({
        data: {
          success: true,
          category: [{ _id: "category-id", name: "Test Category" }],
        },
      });

      let component;
      await act(async () => {
        component = render(<CreateProduct />);
      });

      // Fill in form data for text inputs
      fireEvent.change(screen.getByPlaceholderText("Write a Name"), {
        target: { value: "Test Product" },
      });

      fireEvent.change(screen.getByPlaceholderText("Write a Description"), {
        target: { value: "Test Description" },
      });

      fireEvent.change(screen.getByPlaceholderText("Write a Price"), {
        target: { value: "100" },
      });

      fireEvent.change(screen.getByPlaceholderText("Write a Quantity"), {
        target: { value: "10" },
      });

      // Mock file upload
      const file = new File(["dummy content"], "test.png", {
        type: "image/png",
      });

      const uploadLabel = screen.getByText("Upload Photo");
      const fileInput = uploadLabel.querySelector('input[type="file"]');
      fireEvent.change(fileInput, { target: { files: [file] } });

      // Mock the handleCreate function to avoid Select component interactions
      const createButton = screen.getByText("CREATE PRODUCT");

      fireEvent.click(createButton);

      // Verify API call was made
      await waitFor(() => {
        expect(axios.post).toHaveBeenCalledWith(
          "/api/v1/product/create-product",
          expect.any(Object)
        );
      });

      expect(toast.success).toHaveBeenCalledWith(
        "Product Created Successfully"
      );
      expect(mockedNavigate).toHaveBeenCalledWith("/dashboard/admin/products");
    });

    it("shows error toast when form submission fails", async () => {
      axios.post.mockRejectedValueOnce(new Error("Network Error"));

      await act(async () => {
        render(<CreateProduct />);
      });

      fireEvent.change(screen.getByPlaceholderText("Write a Name"), {
        target: { value: "Test Product" },
      });

      fireEvent.click(screen.getByText("CREATE PRODUCT"));

      await waitFor(() => {
        expect(toast.error).toHaveBeenCalledWith("something went wrong");
      });
    });
  });

  // Snapshot Testing
  describe("Snapshot", () => {
    it("matches snapshot", async () => {
      let renderResult;
      await act(async () => {
        renderResult = render(<CreateProduct />);
      });
      const { container } = renderResult;
      expect(container).toMatchSnapshot();
    });
  });

  // Error Handling Tests
  describe("Error Handling", () => {
    it("handles API errors gracefully", async () => {
      axios.get.mockReset();
      axios.get.mockRejectedValueOnce(new Error("Network Error"));

      await act(async () => {
        render(<CreateProduct />);
      });

      expect(toast.error).toHaveBeenCalledWith(
        "Something went wrong in getting category"
      );

      expect(screen.getByTestId("mock-layout")).toBeInTheDocument();
      expect(screen.getByTestId("mock-adminmenu")).toBeInTheDocument();
    });
  });
});
