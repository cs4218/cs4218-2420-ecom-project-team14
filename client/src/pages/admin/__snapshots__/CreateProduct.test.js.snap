// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateProduct Component Snapshot matches snapshot 1`] = `
<div>
  <div
    data-testid="mock-layout"
    data-title="Dashboard - Create Product"
  >
    <div
      class="container-fluid m-3 p-3"
    >
      <div
        class="row"
      >
        <div
          class="col-md-3"
        >
          <div
            data-testid="mock-adminmenu"
          >
            Admin Menu
          </div>
        </div>
        <div
          class="col-md-9"
        >
          <h1>
            Create Product
          </h1>
          <div
            class="m-1 w-75"
          >
            <div
              class="ant-select ant-select-lg form-select mb-3 css-dev-only-do-not-override-tjsggz ant-select-single ant-select-show-arrow ant-select-show-search"
            >
              <div
                class="ant-select-selector"
              >
                <span
                  class="ant-select-selection-wrap"
                >
                  <span
                    class="ant-select-selection-search"
                  >
                    <input
                      aria-autocomplete="list"
                      aria-controls="rc_select_TEST_OR_SSR_list"
                      aria-expanded="false"
                      aria-haspopup="listbox"
                      aria-owns="rc_select_TEST_OR_SSR_list"
                      autocomplete="off"
                      class="ant-select-selection-search-input"
                      id="rc_select_TEST_OR_SSR"
                      role="combobox"
                      type="search"
                      value=""
                    />
                  </span>
                  <span
                    class="ant-select-selection-placeholder"
                  >
                    Select a Category
                  </span>
                </span>
              </div>
              <span
                aria-hidden="true"
                class="ant-select-arrow"
                style="user-select: none;"
                unselectable="on"
              >
                <span
                  aria-label="down"
                  class="anticon anticon-down ant-select-suffix"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="down"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                    />
                  </svg>
                </span>
              </span>
            </div>
            <div
              class="mb-3"
            >
              <label
                class="btn btn-outline-secondary col-md-12"
              >
                Upload Photo
                <input
                  accept="image/*"
                  hidden=""
                  name="photo"
                  type="file"
                />
              </label>
            </div>
            <div
              class="mb-3"
            />
            <div
              class="mb-3"
            >
              <input
                class="form-control"
                placeholder="Write a Name"
                type="text"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <textarea
                class="form-control"
                placeholder="Write a Description"
                type="text"
              />
            </div>
            <div
              class="mb-3"
            >
              <input
                class="form-control"
                placeholder="Write a Price"
                type="number"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <input
                class="form-control"
                placeholder="Write a Quantity"
                type="number"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <div
                class="ant-select ant-select-lg form-select mb-3 css-dev-only-do-not-override-tjsggz ant-select-single ant-select-show-arrow ant-select-show-search"
              >
                <div
                  class="ant-select-selector"
                >
                  <span
                    class="ant-select-selection-wrap"
                  >
                    <span
                      class="ant-select-selection-search"
                    >
                      <input
                        aria-autocomplete="list"
                        aria-controls="rc_select_TEST_OR_SSR_list"
                        aria-expanded="false"
                        aria-haspopup="listbox"
                        aria-owns="rc_select_TEST_OR_SSR_list"
                        autocomplete="off"
                        class="ant-select-selection-search-input"
                        id="rc_select_TEST_OR_SSR"
                        role="combobox"
                        type="search"
                        value=""
                      />
                    </span>
                    <span
                      class="ant-select-selection-placeholder"
                    >
                      Select Shipping
                    </span>
                  </span>
                </div>
                <span
                  aria-hidden="true"
                  class="ant-select-arrow"
                  style="user-select: none;"
                  unselectable="on"
                >
                  <span
                    aria-label="down"
                    class="anticon anticon-down ant-select-suffix"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="down"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                      />
                    </svg>
                  </span>
                </span>
              </div>
            </div>
            <div
              class="mb-3"
            >
              <button
                class="btn btn-primary"
              >
                CREATE PRODUCT
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
